### PR TITLE
enhancement: more comprehensive filtering in billinglist

### PR DESCRIPTION
**Pull request by interduo**

### CODE REVIEW

File: modules/voipaccountbillinglist.php [hunk 1]
```diff
@@ -103,6 +104,8 @@ function sessionHandler($item, $name)
 unset($voipaccountlist['order']);
 unset($voipaccountlist['direction']);
 
+$voipownerlist = array_column($voipaccountlist, "owner", "ownerid");
```
Comment (chilek) [inline modules/voipaccountbillinglist.php:107]: Użyj Utils::array_column().

**Comment by chilek** [comment]
Poza tym - zmieniasz wybór klienta - lista numerów pozostaje stara, niezależna od wyboru klienta...

**Comment by interduo** [comment]
Nie klikam we frontendy dobrze także zostawię to komuś kto to robi "od strzała". Zanim się zacznę uczyć nowych technologii muszę jeszcze ugasić kilka pożarów w różnych modułach.

File: modules/voipaccountbillinglist.php (hunk 1 above)
Comment (interduo) [reply modules/voipaccountbillinglist.php:107]: Jeśli użyję Utils::array_column() to mam duplikaty i skopiowaną tylko jedną kolumnę voipownerlist.

Jak za pomocą Utils::array_column() uzyskać taki efekt:
```
[4] = klient_z_voipem_id4
[8] = klient_z_voipem_id8

```
?

**Comment by chilek** [comment]
> Zanim się zacznę uczyć nowych technologii muszę jeszcze ugasić kilka pożarów w różnych modułach.

Wczoraj miałeś "ugasić" i co z tego wyszło? :)

File: modules/voipaccountbillinglist.php (hunk 1 above)
Comment (chilek) [reply modules/voipaccountbillinglist.php:107]: Podaj przykład tablicy wejściowej oraz jaki jest oczekiwany efekt dla tablicy wyjściowej. Wówczas będę mógł zweryfikować działanie funkcji Utils::array_column().

**Comment by interduo** [comment]
Pierwszy byłeś.

File: modules/voipaccountbillinglist.php (hunk 1 above)
Comment (interduo) [reply modules/voipaccountbillinglist.php:107]: Wejście:
`Array ( [0] => Array ( [id] => 1 [login] => 814753777 [passwd] => 8jgod9GvbHyMSr38 [ownerid] => 7 [owner] => ASDASDADS 123 [access] => 1 [borough_name] => [district_name] => [state_name] => [city_name] => [street_name] => [street_type] => [location_name] => [location_city_name] => Lubartów [location_street_name] => ul. Lubelska 36B [location_city] => [location_street] => [location_house] => [location_flat] => [location] => 21-100 Lubartów, ul. Lubelska 36B [phone] => Array ( [0] => 814753777 ) ) [1] => Array ( [id] => 3 [login] => 814753779 [passwd] => mK3qY8sIhWHh8HHo [ownerid] => 3 [owner] => DDDD asdasd [access] => 1 [borough_name] => [district_name] => [state_name] => [city_name] => [street_name] => [street_type] => [location_name] => [location_city_name] => SSS [location_street_name] => [location_city] => [location_street] => [location_house] => [location_flat] => [location] => SSS, SSS [phone] => Array ( [0] => 814753779 ) ) [2] => Array ( [id] => 4 [login] => 814753780 [passwd] => cuqHRH5Jt6h6Cv5W [ownerid] => 4 [owner] => KOWALSKI Jane [access] => 1 [borough_name] => [district_name] => [state_name] => [city_name] => [street_name] => [street_type] => [location_name] => [location_city_name] => Lubartów [location_street_name] => ul. Lubelska 36B [location_city] => [location_street] => [location_house] => [location_flat] => [location] => 21-100 Lubartów, ul. Lubelska 36B [phone] => Array ( [0] => 814753780 ) ) [3] => Array ( [id] => 5 [login] => 814753781 [passwd] => D8zMIMujF0ytA0u4 [ownerid] => 8 [owner] => MALINOWSKI Jan [access] => 1 [borough_name] => [district_name] => [state_name] => [city_name] => [street_name] => [street_type] => [location_name] => [location_city_name] => Lubartów [location_street_name] => [location_city] => [location_street] => [location_house] => [location_flat] => [location] => Lubartów, Lubartów [phone] => Array ( [0] => 814753781 ) ) [4] => Array ( [id] => 2 [login] => 814753778 [passwd] => 8jgod9GvbHyMSr38 [ownerid] => 2 [owner] => TEST TEST [access] => 1 [borough_name] => Lubartów [district_name] => lubartowski [state_name] => lubelskie [city_name] => Lubartów [street_name] => Lubelska [street_type] => ul. [location_name] => [location_city_name] => Lubartów [location_street_name] => ul. Lubelska [location_city] => 12617 [location_street] => 11393 [location_house] => 22 [location_flat] => [location] => 21-100 Lubartów, ul. Lubelska 22 [phone] => Array ( [0] => 814753778 ) ) [5] => Array ( [id] => 6 [login] => 814753783 [passwd] => 0AXLMSjJ5lvNuyUF [ownerid] => 9 [owner] => TRUSKAWSKI Krzysztof [access] => 1 [borough_name] => [district_name] => [state_name] => [city_name] => [street_name] => [street_type] => [location_name] => [location_city_name] => Lubartów [location_street_name] => [location_city] => [location_street] => [location_house] => [location_flat] => [location] => Lubartów, Lubartów [phone] => Array ( [0] => 814753783 ) ) [6] => Array ( [id] => 7 [login] => 814753784 [passwd] => 0AXLMSjJ5lvNuyUF [ownerid] => 9 [owner] => TRUSKAWSKI Krzysztof [access] => 1 [borough_name] => [district_name] => [state_name] => [city_name] => [street_name] => [street_type] => [location_name] => [location_city_name] => Lubartów [location_street_name] => [location_city] => [location_street] => [location_house] => [location_flat] => [location] => Lubartów, Lubartów [phone] => Array ( [0] => 814753784 ) ) )`


Oczekiwane (ownerid => ownername):
`Array ( [7] => ASDASDADS 123 [3] => DDDD asdasd [4] => KOWALSKI Jane [8] => MALINOWSKI Jan [2] => TEST TEST [9] => TRUSKAWSKI Krzysztof )`

Comment (chilek) [reply modules/voipaccountbillinglist.php:107]: Proszę: https://github.com/lmsgit/lms/commit/6e3880b94d591b6ab17fe8a4d52a5800ec0e2e9c

Comment (interduo) [reply modules/voipaccountbillinglist.php]: Dzięki. A tak btw jaki jest cel tego by nie używać tej funkcji bezpośrednio?

Comment (chilek) [reply modules/voipaccountbillinglist.php]: Minimalizacja odchyleń od zgodności z PHP 5.4.

**Comment by alecpl** [comment]
https://github.com/ramsey/array_column

File: modules/voipaccountbillinglist.php (hunk 1 above)
Comment (interduo) [reply modules/voipaccountbillinglist.php]: Dzięki za wyjaśnienie.

**Comment by interduo** [comment]
Zapewne jestem blisko oczekiwanego efektu ale nie wiem gdzie błąd bo nie umiem tego zdebugować.

**Comment by interduo** [comment]
> Poza tym - zmieniasz wybór klienta - lista numerów pozostaje stara, niezależna od wyboru klienta...

Zrobione.

File: templates/default/voipaccount/voipaccountbillinglist.html [hunk 2]
```diff
@@ -109,13 +109,16 @@ <h1>{$layout.pagetitle}</h1>
...
-						<option value="{$vacc.id}" {if $listdata.fvoipaccid == $vacc.id} selected{/if}>{$vacc.owner|truncate:40:"&hellip;":true} ({$vacc.login})</option>
-						{/foreach}
-					</select>
+                                       {trans("Customer:")}
+                                       <select name="fvownerid" class="lms-ui-advanced-select" onChange="CustomerChanged($(this).val())">
+					       <option value="" {if !$listdata.fvownerid}selected{/if}>{trans("- all -")}</option>
+                                               {foreach from=$voipownerlist item=vown key=$idx}
+                                               <option value="{$idx}" {if $listdata.fvownerid == $idx} selected{/if}>{$vown|truncate:40:"...":true}</option>
+                                                {/foreach}
+                                       </select>
```
Comment (chilek) [inline templates/default/voipaccount/voipaccountbillinglist.html:118]: Co nietegez z formatowaniem - {/foreach} przesunięty?

File: templates/default/voipaccount/voipaccountbillinglist.html [hunk 3]
```diff
@@ -318,13 +322,19 @@ <h1>{$layout.pagetitle}</h1>
...
-			$('#choosefilter').submit();
+	function CustomerChanged(str) {
+		$.ajax({
+			url: '?m=voipaccountbillinglist',
+			type: "POST",
+			data: { str : str },
+			success:
+				function(response) {
+					var selectbox = $('#customernumberselect');
+					var parsed = jQuery.parseJSON( response );
```
Comment (chilek) [inline templates/default/voipaccount/voipaccountbillinglist.html:333]: Czego tu oczkujesz w formacie JSON? Od razu przy $.ajax() możesz poinformować jQuery, że oczekujesz odpowiedzi JSON i otrzymasz od razu sparsowaną. Zobacz w innych miejscach jak to robione.

**Review by chilek — CHANGES_REQUESTED** [review]
Jak w komentarzach.

File: templates/default/voipaccount/voipaccountbillinglist.html (hunk 3 above)
Comment (interduo) [reply templates/default/voipaccount/voipaccountbillinglist.html]: Faktycznie kod wygląda ładniej.

https://github.com/lmsgit/lms/pull/1819/commits/22c02c5c18acaec7b7a1a867eb9b27e5a5a1a645

File: templates/default/voipaccount/voipaccountbillinglist.html (hunk 2 above)
Comment (interduo) [reply templates/default/voipaccount/voipaccountbillinglist.html]: https://github.com/lmsgit/lms/pull/1819/commits/1d8644a316db9d6ac6b141e42f6114d1bbd6f592

**Comment by interduo** [comment]
Poszły poprawki i testy już na zielono.

**Comment by interduo** [comment]
@chilek coś temu brakuje? My tego używamy już na produkcji już jakiś miesiąc. Zrobiłem rebase żebyś mógł łatwiej wrzucić cherry-pickiem sobie do testów.

**Comment by interduo** [comment]
@chilek  ?

**Comment by interduo** [comment]
@chilek ?

File: templates/default/voipaccount/voipaccountbillinglist.html [hunk 4]
```diff
@@ -110,13 +110,16 @@ <h1>{$layout.pagetitle}</h1>
...
+						<option value="" {if !$listdata.fvownerid}selected{/if}>{trans("- all -")}</option>
+						{foreach from=$voipownerlist item=vown key=$idx}
+						<option value="{$idx}" {if $listdata.fvownerid == $idx} selected{/if}>{$vown|truncate:40:"...":true}</option>
 						{/foreach}
 					</select>
 
+                                       {trans("Number:")}
+                                       {include file="voipaccount/voipaccountnumbers.html"}
+
 					{trans("Period")} {trans("from")}&nbsp;
```
Comment (chilek) [inline templates/default/voipaccount/voipaccountbillinglist.html:123]: Dlaczego tu formatowanie więc inne niż w reszcie szablonu?

**Comment by chilek** [comment]
10q

**Comment by chilek** [comment]
Podejrzewam, że to rozwali przejście do billingów z konta VoIP.
Filtrowanie powinno być 3-poziomowe:
1. Klient.
2. Konta VoIP przypisane do klienta.
3. Numery przypisane do konta VoIP wybranego w punkcie 2 lub dla wszystkich kont wybranego klienta.

**Comment by interduo** [comment]
> Filtrowanie powinno być 3-poziomowe

Proponuję rozbudujmy najpierw o ten drugi poziom a potem o kolejny jeśli ktoś wyrazi taką potrzebę.

File: templates/default/voipaccount/voipaccountbillinglist.html (hunk 4 above)
Comment (interduo) [reply templates/default/voipaccount/voipaccountbillinglist.html:123]: https://github.com/lmsgit/lms/pull/1819/commits/67df5efe1fb9ef40eda0f91eb51cf18b1246f4f1

miałem poprawić ale mnie ubiegłeś, a może hurtem przy kolejnym wydaniu zmienimy taby na spacje?

**Comment by chilek** [comment]
Poproszę o zrzut ekranu demonstrujący jak to wygląda obecnie oraz jak po zmianie?

**Comment by interduo** [comment]
Filtry PRZED:
Zrzut ekranu z 2021-03-10 15-10-21

Filtry PO:
Zrzut ekranu z 2021-03-10 15-09-38

**Comment by chilek** [comment]
To od razu mam zastrzeżenie - wybór klienta powinien opierać się na elemencie smarty `{customerlist ...}`.
W ogóle nie ma oczym gadać...

**Comment by interduo** [comment]
@chilek wymieniłem selecta na {customerlist

**Comment by chilek** [comment]
Poproszę o zrzuty ekranu jak to wygląda obecnie?

**Comment by interduo** [comment]
filtry

**Comment by chilek** [comment]
Dlaczego następuje łamanie wiersza filtru po wyborze klienta? Nie przypominam sobie, żeby w innych miejsach tak było?

**Comment by interduo** [comment]
Nie mam pojęcia ale w sumie to nie jest złe bo dzięki temu nie będzie pływania UI.

**Comment by interduo** [comment]
A o tym co sądzisz?

**Comment by chilek** [comment]
Wybór klienta i od razu wybór numeru - brakuje po środku wyboru konta VoIP.

**Comment by interduo** [comment]
> Wybór klienta i od razu wybór numeru - brakuje po środku wyboru konta VoIP.

To już kolejny filtr.

W kwestii filtrowania jeszcze przydałby się filtry:
- select grupy prefiksów - tak by móc np. wyfiltrować połączenia płatne dla danego numeru,
- filtr numer telefonu - możliwość wyszukania ręcznie wpisanego numeru jako CPA lub CPB. 

No a ma być małymi zmianami.

**Comment by interduo** [comment]
Może puścisz już ten gotowy zakres i zrobimy issue na te pozostałe rzeczy?

**Comment by chilek** [comment]
> Może puścisz już ten gotowy zakres i zrobimy issue na te pozostałe rzeczy?

Myślałem, że zrobimy od razu docelowo i spróbujemy puścić do 26.x.
W obecnej postaci to jest pewnego rodzaju "regresja" i nie za bardzo kwalifikuje się do 26.x :(

**Comment by interduo** [comment]
Podszedłem do tematu inaczej podzieliłem obecny filtr na filtr klient i filtr konto VoIP. 
Jest to wystarczające. Jak będzie kiedyś potrzeba wybrać filtr "numer" to dorobimy.  
Ja takiej wyraźnej potrzeby nie mam.

**Comment by chilek** [comment]
> Podszedłem do tematu inaczej podzieliłem obecny filtr na filtr klient i filtr konto VoIP. Jest to wystarczające. Jak będzie kiedyś potrzeba wybrać filtr "numer" to dorobimy. Ja takiej wyraźnej potrzeby nie mam a nie chcę robić prowizorki. Nie ma już regresu.

Filtr numeru jest używany przez parę osób. Co mam im powiedzieć? :D

**Comment by interduo** [comment]
> Filtr numeru jest używany przez parę osób. Co mam im powiedzieć? :D

Powiedz im, że to nie jest filtr "numeru" a złączony filtr "Klient (LoginKontaVoIP).

**Comment by chilek** [comment]
> To jest wystarczający dowód na to że większość osób używa mapowania 1:1, jeden numer = jeden login.

W przypadku naszego wdrożenia u operatora, który hurtem sprzedaje usługi telefoniczne akurat dużo numerów przypada na konto i wiele kont na klienta.

**Comment by chilek** [comment]
W sumie do wyboru numeru i tak nie było, więc to co zrobiłeś to jest zrobienie z wyboru jednopoziomowego wyboru dwu poziomowego!

**Comment by chilek** [comment]
10q
Wpis w changelog dałeś?

**Comment by interduo** [comment]
> W przypadku naszego wdrożenia u operatora, który hurtem sprzedaje usługi telefoniczne akurat dużo numerów przypada na konto i wiele kont na klienta.

No tak - zgadza się. My planujemy tak dodawać tak SIP Trunki dla klientów biznesowych. Ale co innego miałem na myśli wyżej :)

> Wpis w changelog dałeś?

https://github.com/lmsgit/lms/pull/1972

**Comment by interduo** [comment]
A co Ty na to, żeby przestać korzystać z doc/Changelog na korzyść mechanizmu z githuba?
Ewentualnie generować go automatycznie z githuba? 
Uważam, że trzeba zrezygnować takich małpich prac - tymbardziej z perspektywy czasu mogę powiedzieć że to jest najbardziej konfliktodajny plik.

**Comment by chilek** [comment]
> A co Ty na to, żeby przestać korzystać z doc/Changelog na korzyść mechanizmu z githuba?

Niewykonalne, bo w change opisujesz istotne rzeczy, a commitów generujesz mnóstwo - np. każde wciśnięcie literki u Ciebie w klawiaturze :)

**Comment by chilek** [comment]
Zauważ, że wszystkie szanujące się projektu utrzymują plik ChangeLog.

**Comment by interduo** [comment]
To może `| egrep -i (enhancement|improvement)` ?

**Comment by interduo** [comment]
Albo jeszcze lepiej:

`git log --pretty=format:'%C(yellow)%h %Cred%ad %Cblue%an%Cgreen%d %Creset%s' --date=short`

**Comment by chilek** [comment]
Zobacz jak robią inni z wieloletnim doświadczeniem. Można również poradzić się naszego mentora, @alecpl ? :)

**Comment by interduo** [comment]
Chyba przyznasz, że papierowy changelog w postaci pliku txt to naleciałość jeszcze projektów tworzonych przed użyciem systememów kontroli wersji.

Zauważ, że teraz wiele projektów robi np. tak: wstrzela commita z msg np. "released version 26.23" w każdym z branchu i ma temat utrzymywania changeloga z głowy. Widać wtedy z wersji na wersję numerowaną co się zmieniło. 

Argumenty za:
- nie musisz już kopiować commitmsg do changeloga,
- widzisz zmiany między numerkami nie tylko między seriami,
- nie musisz poprawiać konfliktów

> Zobacz jak robią inni z wieloletnim doświadczeniem.

Zrobiłem już research już dawno. Zobacz jak jest w nowych projektach ludzi, którzy mają świeże spojrzenie - jest to robione w drugą stronę takimi narzędziami (jest ich kupa) np. https://github.com/github-changelog-generator/github-changelog-generator#why-should-i-care

A z chęcią posłucham jak @alecpl by zrobił gdyby od nowa tworzył jakiś projekt.

**Comment by chilek** [comment]
W dniu 17.03.2021 10:15, Interduo napisał(a):
> Chyba przyznasz, że papierowy changelog w postaci pliku txt to
> naleciałość jeszcze projektów tworzonych przed użyciem
> systememów kontroli wersji.
> 
> Zauważ, że teraz wiele projektów robi np. tak: wstrzela commita z
> msg np. "released version 26.23" w każdym z branchu i ma temat
> utrzymywania changeloga z głowy. Widać wtedy z wersji na wersję
> numerowaną co się zmieniło.
> 
> Argumenty za:
> 
>  	* nie musisz już kopiować commitmsg do changeloga,
>  	* widzisz zmiany między numerkami nie tylko między seriami,
>  	* nie musisz poprawiać konfliktów
> 
>> Zobacz jak robią inni z wieloletnim doświadczeniem.
> 
> Zrobiłem już research już dawno. Zobacz jak jest w nowych
> projektach ludzi, którzy mają świeże spojrzenie - jest to robione
> w drugą stronę takimi narzędziami (jest ich kupa) np.
> https://github.com/github-changelog-generator/github-changelog-generator#why-should-i-care

Jeśli byśmy prefiksowali odpowiedni fragment commit message-ów to można 
byłoby oczywiście tworzyć
automatycznie ChangeLog, ale - pamiętaj, że z commit-messages problem 
jest taki, że nie da się
ich poprawić w historii!!

**Comment by interduo** [comment]
Czy kiedyś poprawialiśmy commit messages?
Dla najnowszego commita można to jak najbardziej zrobić za pomocą rebase oraz opcji reword.

**Comment by chilek** [comment]
> Czy kiedyś poprawialiśmy commit messages w zamierzchłej przeszłości?

Parę razy mi się zdarzyło, gdy był rażący błąd w tekście, ale zgadnij dlaczego zwykle nie poprawiamy?
Właśnie dlatego, że nie da się poprawić w historii bez późniejszego git push z plusem.

**Comment by interduo** [comment]
Czy to nie problem z cyklu wydarzy się raz w kolejnych dwóch-trzech latach?:)

**Comment by chilek** [comment]
> Czy to nie problem z cyklu wydarzy się raz w kolejnych dwóch-trzech latach?:)

No nie wiem - po Tobie ChangeLog trzeba często poprawiać. Prawdę pisząc właściwie każdy Twój wpis jest do poprawy ;-)

**Comment by alecpl** [comment]
Automatyczne generowanie changeloga wymaga dyscypliny, a tego nie lubię ;) Nie mam w tym zakresie zbyt dużego  doświadczenia.
https://github.com/github-changelog-generator/github-changelog-generator powołuje się na https://keepachangelog.com ale sam robi dokładnie to czego ta strona nie doradza, czyli "Don’t let your friends dump git logs into changelogs". Na przykład 
```
Merged pull requests:
    - Tell the truth in this comment #792 (olleolleolle)
```
To naprawdę dużo mówiący wpis ;)

**Comment by chilek** [comment]
Hehe, można jeszcze w commit message rozważyć wprowadzanie fragmentów typu:
CHANGELOG: opis zmian coś mówiący
ale znowu w historii nie będzie prosto poprawić literówek np. po @interduo.

Poza tym jak często dodajemy wpisy do ChangeLog, żeby prowadzić debatę w zakresie optymalizacji tego? :D

**Comment by interduo** [comment]
Z drugiej strony jest jeszcze jeden problem - obecna forma dla użytkowników końcowych jest mało zrozumiała.
Traktujemy to jako ChangeLog dla programisty/admina, prawda? Te osoby na bank umieją w git log

**Comment by chilek** [comment]
> Z drugiej strony jest jeszcze jeden problem - obecna forma dla użytkowników końcowych jest mało zrozumiała.

Nie uważam tak - staramy się w ChangeLog opisywać bardziej przyjaźnie niż w commit message. Tylko od nas zależy
w jakim stopniu jest coś tu zrozumiałe.

**Comment by interduo** [comment]
Zgadzam się, że się staramy. Zdecydowanie przyjaźniejszą formą byłby artykuł na wiki ze zrzutami przed i po. Ewentualnie demo funkcji w postaci wideo. Wyślij to do pracowników mniej technicznych biura/administracyjnych to zobaczysz o czym mówię.

**Pull request merged**